apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs

```diff
@@ -467,5 +498,95 @@ private void CheckNotDisposed()
                 throw new ObjectDisposedException($"PooledSession for {SessionName} has been disposed, and cannot be reused.");
             }
         }
+
+        /// <summary>
+        /// Executes the <paramref name="asyncWork"/> with appropritae transaction selector.
+        /// </summary>
+        /// <typeparam name="TResult"> The type of result to be retunred.</typeparam>
+        /// <param name="asyncWork"> Async task that will assign appropriate transaction selector
+        /// to the request and then execute the request.</param>
+        /// <param name="cancellationToken"> The cancellation token.</param>
+        /// <param name="currentTransactionSelector"> The current transaction selector of request that needs to be updated.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        public async Task<TResult> ExecuteWithTransactionSelector<TResult>(Func<TransactionSelector, Task<TResult>> asyncWork, TransactionSelector currentTransactionSelector, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNull(asyncWork, nameof(asyncWork));
+
+            // If transaction options is null it means tramsaction is not required
+            // to execute this operation. Which means It should have been invoked from Read Only Transaction.
+            // If current tramsaction selector if request is of type single use, the no need to inline begin a new transaction.
+            if (TransactionOptions == null ||
+                currentTransactionSelector?.SelectorCase == TransactionSelector.SelectorOneofCase.SingleUse)
+            {
+                return await asyncWork(currentTransactionSelector).ConfigureAwait(false);
+            }
+
+            // Transaction has been already started by an earlier operation.
+            if (TransactionId != null)
+            {
+                return await asyncWork(new TransactionSelector { Id = TransactionId }).ConfigureAwait(false);
+            }
+
+            // Null Transaction Id means a new trasaction needs to inline begin/started.
+            // Only one thread should inline begin the transction.
+
+            var enteredSemaphoreSlim = false; // Intially false for all threads.
+            if (TransactionId == null)
+            {
+                await _semaphoreSlim.WaitAsync().ConfigureAwait(false); // All except one thrad will wait here for semaphore to release.
+                enteredSemaphoreSlim = true; // True for first thread trying to inline begin.
+            }
```

Comment:
I don't think this works as you think it does:

Say 5 commands are all executed at "the same time". For all of them TransactionId will be null so all of them will attempt to grab the semaphore. Only one of them will grab the semaphore.

That one will execute and inline the transaction. In the meantime, 3 more commands were executed, and for all of those TransactionId was null because the first command that inlined the transaction is not done yet. You now have 7 commands waiting for the semaphore.

Now th first command returns, sets TransactionId to some value and releases the semaphore. One of the 7 commands waiting for the semaphore grabs it, while the other 6 continue to be blocked there.

That one command that grabbed the semaphore gets to execute with the new transaction Id and after execution, releases the semaphore, so one of the 6 remaining gets to execute, and then one of the 5 remaining, etc.

The problem is that after the first command that inlined the transaction returns, all 7 waiting commandas can start executing in parallel, and this implementation doesn't allow that.

Reply:
Yes, that's true for all the commands executed before TransactionId generation. 
So the scenario you have explained, to handle that we can release semaphore from the thread as soon as transactionId is not null for next threads.
That is adding if block at Line number 541 like below:
```
// Transaction was aready inline begin by another thread in parallel while this was on hold. 
            if (TransactionId != null)
            {
             if (enteredSemaphoreSlim)
                {
                    _semaphoreSlim.Release();
                    enteredSemaphoreSlim=false;
                }
                return await asyncWork(new TransactionSelector { Id = TransactionId }).ConfigureAwait(false);
            }
            ```

Reply:
Do you think it'll work ?

Reply:
Not really. Yes, the blocked commands don't have to wait for the other ex-blocked commands to finish executing, but still, they have to wait to acquiere the semaphore. Basically, we'll have waiting commands that don't need to be waiting, and that's not good even if for a small amount of time.